### ISSUE TITLE
Failed writing data to layaway file: no _dump_data is defined for class Thread::Mutex`

### ISSUE THREAD

**Issue by bjeanes** [scoutapp/scout_apm_ruby#211]
In upgrading a Rails application from Rails 4.1 to 4.2, Scout stops sending data entirely. The logs contain (I've formatted the backtrace-as-array that is printed):

```
Failed writing data to layaway file: no _dump_data is defined for class Thread::Mutex / 
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/layaway_file.rb:32:in `dump'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/layaway_file.rb:32:in `serialize'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/layaway_file.rb:27:in `write'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/layaway.rb:61:in `write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:102:in `block in write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:101:in `synchronize'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:101:in `write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:91:in `block in write_to_layaway'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:91:in `each'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/store.rb:91:in `write_to_layaway'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/reporting.rb:31:in `process_metrics'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/periodic_work.rb:13:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/agent.rb:173:in `block (2 levels) in start_background_worker'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/background_worker.rb:61:in `block in start'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/background_worker.rb:44:in `loop'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/background_worker.rb:44:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-2.4.14/lib/scout_apm/agent.rb:172:in `block in start_background_worker'
```

I tried upgrading to pre-release v3 gem to see if it changes, but same thing (and something else):

```
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/trace_compactor.rb:201: [BUG] frame2iseq: unreachable
7efe53b92000-7efe53b96000 r-xp 00000000 fc:01 5374097 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/stacks.so
7efe53b96000-7efe53d95000 ---p 00004000 fc:01 5374097 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/stacks.so
7efe53d95000-7efe53d96000 r--p 00003000 fc:01 5374097 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/stacks.so
7efe53d96000-7efe53d97000 rw-p 00004000 fc:01 5374097 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/stacks.so
7efe53d97000-7efe53d98000 r-xp 00000000 fc:01 5374099 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/allocations.so
7efe53d98000-7efe53f98000 ---p 00001000 fc:01 5374099 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/allocations.so
7efe53f98000-7efe53f99000 r--p 00001000 fc:01 5374099 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/allocations.so
7efe53f99000-7efe53f9a000 rw-p 00002000 fc:01 5374099 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/allocations.so
7efe53f9a000-7efe53f9c000 r-xp 00000000 fc:01 5374102 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/rusage.so
7efe53f9c000-7efe5419b000 ---p 00002000 fc:01 5374102 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/rusage.so
7efe5419b000-7efe5419c000 r--p 00001000 fc:01 5374102 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/rusage.so
7efe5419c000-7efe5419d000 rw-p 00002000 fc:01 5374102 /app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/rusage.so
Failed writing data to layaway file: no _dump_data is defined for class Thread::Mutex /
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/layaway_file.rb:32:in `dump'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/layaway_file.rb:32:in `serialize'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/layaway_file.rb:27:in `write'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/layaway.rb:61:in `write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:102:in `block in write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:101:in `synchronize'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:101:in `write_reporting_period'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:91:in `block in write_to_layaway'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:91:in `each'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/store.rb:91:in `write_to_layaway'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/reporting.rb:31:in `process_metrics'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/periodic_work.rb:13:in `run'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/agent.rb:179:in `block (2 levels) in start_background_worker'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/background_worker.rb:61:in `block in start'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/background_worker.rb:44:in `loop'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/background_worker.rb:44:in `start'
/app/vendor/bundle/ruby/2.3.0/gems/scout_apm-3.0.0.pre25/lib/scout_apm/agent.rb:178:in `block in start_background_worker'
```

**Comment by bjeanes** [comment]
I diffed the Gemfiles for the different deployments and post-processed it a bit to show old version and new versions:

Rails 4.1                              | Rails 4.2
----------                             | ----------
_n/a_                                  | crass (1.0.4)
_n/a_                                  | globalid (0.4.1)
_n/a_                                  | loofah (2.2.2)
_n/a_                                  | rails-deprecated_sanitizer (1.0.3)
_n/a_                                  | rails-dom-testing (1.0.9)
_n/a_                                  | rails-html-sanitizer (1.0.4)
_n/a_                                  | activejob (4.2.10)
actionmailer (4.1.16)                  | actionmailer (4.2.10)
actionpack (4.1.16)                    | actionpack (4.2.10)
actionview (4.1.16)                    | actionview (4.2.10)
activemodel (4.1.16)                   | activemodel (4.2.10)
activerecord (4.1.16)                  | activerecord (4.2.10)
activerecord-postgresql-citext         | _n/a_
activerecord-postgresql-citext (0.2.0) | _n/a_
activesupport (4.1.16)                 | activesupport (4.2.10)
arel (5.0.1.20140414130214)            | arel (6.0.4)
has_scope (0.6.0)                      | has_scope (0.7.2)
inherited_resources (1.7.2)            | inherited_resources (1.8.0)
rack (1.5.5)                           | rack (1.6.10)
rails (4.1.16)                         | rails (4.2.10)
rails (< 4.2)                          | rails (~> 4.2)
railties (4.1.16)                      | railties (4.2.10)
responders (1.1.2)                     | responders (2.4.0)
sprockets-rails (2.3.3)                | sprockets-rails (3.2.1)

**Comment by dlanderson** [comment]
Hi @bjeanes - Do you happen to have Datadog installed? We've seen one case of this before where a datadog thread caused issue with Scout.

If you don't have Datadog installed we have a debug branch of Scout that can gather more information for us.

**Comment by bjeanes** [comment]
Yes, I do! Is there a solution to running them side by side? I don't want
to sacrifice either right now.

**Comment by bjeanes** [comment]
I spent some time tonight with a debugger after replicating the issue locally.

I found it very hard to lock down the data long enough to inspect it so this took a while, but I eventually found where the Datadog objects were being stored in the data that is undergoing dumping.

I only saw this in slow transaction traces, but the `@extra` ivar on `ScoutApm::MetricMeta` sometimes ends up with a `:datadog_span` key, which contains a HUGE hierarchy of DD objects, including some `Thread::Mutex`es.:

<img width="1044" alt="2018-08-04 at 11 51 pm" src="https://user-images.githubusercontent.com/2560/43677239-e6d12e96-9841-11e8-9afe-7b5e1a4d3ec0.png">

As best I can tell, Datadog _mutates_ the `payload` passed around through `ActiveSupport::Notification` events here https://github.com/DataDog/dd-trace-rb/blob/40024f8f36a5d65e14f1c3bdd318f833d12cd450/lib/ddtrace/contrib/active_support/notifications/subscription.rb#L74-L79 and then Scout annotates the current layer with the entire payload https://github.com/scoutapp/scout_apm_ruby/blob/5e9ab893db185faca3d596e45e93ffa3497904a3/lib/scout_apm/instruments/active_record.rb#L141-L151

I wonder if Scout could be more selective about what keys it expects to see in that payload (or what keys it actually does anything with). Alternatively, it could explicitly remove the `:datadog_span` key from consideration.

Or, perhaps the marshalling spec can be more defensive and remove certain objects that can't be marshalled (procs, mutexes, etc) from the hierarchy altogether.

Confusingly, the `data` object being dumped also has a reference to the Scout context, which also contains a `Thread::Mutex` but that one doesn't seem to cause an issue, but I don't understand how it doesn't since there's no obvious pruning of that data before passing it directly to `Marshal.dump`.

**Comment by dlanderson** [comment]
Closed in #212

**Comment by dlanderson** [comment]
Fixed in 2.4.15 (.pre26 for 3.0 versions)

**Comment by bjeanes** [comment]
You rock! Thanks for releasing this so promptly :)
